SofaKernel/modules/SofaBaseMechanics/BarycentricMappers/BarycentricMapperTopologyContainer.inl

```diff
@@ -146,94 +147,107 @@ template <class In, class Out, class MappingDataType, class Element>
     {
-        Vec3d outPos = Out::getCPos(out[i]);
-        Vector3 baryCoords;
-        int elementIndex = -1;
-        double distance = std::numeric_limits<double>::max();
+        Vector3 outPos = Out::getCPos(out[i]);
+        NearestParams nearestParams;
 
+        // Search nearest element in grid cell
+        Vec3i gridIds = getGridIndices(outPos);
         unsigned int h = getHashIndexFromCoord(outPos);
-        for ( unsigned int j=0; j<m_hashTable[h].size(); j++)
+        for (unsigned int j=0; j<m_hashTable[h].size(); j++)
         {
-            unsigned int e = m_hashTable[h][j];
-            Vec3d bary = bases[e] * ( outPos - in[elements[e][0]] );
-            double dist;
-            computeDistance(dist, bary);
-            if ( dist>0 )
-                dist = ( outPos-centers[e] ).norm2();
-            if ( dist<distance )
+            HashEntry entry = m_hashTable[h][j];
+            if(sameGridCell(entry, gridIds))
             {
-                baryCoords = bary;
-                distance = dist;
-                elementIndex = int(e);
+                unsigned int e = entry.elementId;
+                Vector3 inPos = in[elements[e][0]];
+                checkDistanceFromElement(e, outPos, inPos, nearestParams);
             }
         }
 
-        if(elementIndex==-1)
+        if(nearestParams.elementIndex==-1) // No element in grid cell, perform exhaustive search
+        {
+            for ( unsigned int e = 0; e < elements.size(); e++ )
+            {
+                Vector3 inPos = in[elements[e][0]];
+                checkDistanceFromElement(e, outPos, inPos, nearestParams);
+            }
+        }
+        else if(abs(nearestParams.distance)>m_gridCellSize/2.) // Nearest element in grid cell may not be optimal, check neighbors
         {
-            exhaustiveSearch(outPos, in, bases, centers);
+            Vec3i centerGridIds = gridIds;
+            for(int xId=-1; xId<=1; xId++)
+                for(int yId=-1; yId<=1; yId++)
+                    for(int zId=-1; zId<=1; zId++)
+                    {
+                        gridIds = Vec3i(centerGridIds[0]+xId,centerGridIds[1]+yId,centerGridIds[2]+zId);
+                        h = getHashIndexFromIndices(gridIds);
+                        for ( unsigned int j=0; j<m_hashTable[h].size(); j++)
+                        {
+                            HashEntry entry = m_hashTable[h][j];
+                            if(sameGridCell(entry, gridIds))
+                            {
+                                unsigned int e = entry.elementId;
+                                Vector3 inPos = in[elements[e][0]];
+                                checkDistanceFromElement(e, outPos, inPos, nearestParams);
+                            }
+                        }
+                    }
         }
-        else
-            addPointInElement(elementIndex, baryCoords.ptr());
+
+        addPointInElement(nearestParams.elementIndex, nearestParams.baryCoords.ptr());
```

Comment:
Sorry I didn't spent time to understand your algorithm. Just looking quickly at the "if" and "else if" conditions could there be an "else" case ?
If yes but should not happen, maybe add a msg_error/warning

Reply:
no it is ok, leave it as is. As I said I didn't spent time to understand the code, just saw if/else if and not else.